// amadeus/lib/amadeus/client/errors.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.UnknownError = exports.ServerError = exports.ResponseError = exports.ParserError = exports.NotFoundError = exports.NetworkError = exports.ClientError = exports.AuthenticationError = void 0;
function _typeof(o) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (o) { return typeof o; } : function (o) { return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o; }, _typeof(o); }
function _callSuper(t, o, e) { return o = _getPrototypeOf(o), _possibleConstructorReturn(t, _isNativeReflectConstruct() ? Reflect.construct(o, e || [], _getPrototypeOf(t).constructor) : o.apply(t, e)); }
function _possibleConstructorReturn(t, e) { if (e && ("object" == _typeof(e) || "function" == typeof e)) return e; if (void 0 !== e) throw new TypeError("Derived constructors may only return object or undefined"); return _assertThisInitialized(t); }
function _assertThisInitialized(e) { if (void 0 === e) throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); return e; }
function _isNativeReflectConstruct() { try { var t = !Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); } catch (t) {} return (_isNativeReflectConstruct = function _isNativeReflectConstruct() { return !!t; })(); }
function _getPrototypeOf(t) { return _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf.bind() : function (t) { return t.__proto__ || Object.getPrototypeOf(t); }, _getPrototypeOf(t); }
function _inherits(t, e) { if ("function" != typeof e && null !== e) throw new TypeError("Super expression must either be null or a function"); t.prototype = Object.create(e && e.prototype, { constructor: { value: t, writable: !0, configurable: !0 } }), Object.defineProperty(t, "prototype", { writable: !1 }), e && _setPrototypeOf(t, e); }
function _setPrototypeOf(t, e) { return _setPrototypeOf = Object.setPrototypeOf ? Object.setPrototypeOf.bind() : function (t, e) { return t.__proto__ = e, t; }, _setPrototypeOf(t, e); }
function _classCallCheck(a, n) { if (!(a instanceof n)) throw new TypeError("Cannot call a class as a function"); }
function _defineProperties(e, r) { for (var t = 0; t < r.length; t++) { var o = r[t]; o.enumerable = o.enumerable || !1, o.configurable = !0, "value" in o && (o.writable = !0), Object.defineProperty(e, _toPropertyKey(o.key), o); } }
function _createClass(e, r, t) { return r && _defineProperties(e.prototype, r), t && _defineProperties(e, t), Object.defineProperty(e, "prototype", { writable: !1 }), e; }
function _toPropertyKey(t) { var i = _toPrimitive(t, "string"); return "symbol" == _typeof(i) ? i : i + ""; }
function _toPrimitive(t, r) { if ("object" != _typeof(t) || !t) return t; var e = t[Symbol.toPrimitive]; if (void 0 !== e) { var i = e.call(t, r || "default"); if ("object" != _typeof(i)) return i; throw new TypeError("@@toPrimitive must return a primitive value."); } return ("string" === r ? String : Number)(t); }
/**
 * The error that is passed to the Promise when the API call fails.
 *
 * @param {Response} response the {@link Response} object containing the raw
 *  http response and the {@link Request} instance that made the API call.
 * @property {Response} response the {@link Response} object containing the raw
 *  http response and the {@link Request} instance that made the API call.
 * @property {string} code a unique code for this type of error. Options include
 *  `NetworkError`, `ParserError`, `ResponseError`, `ServerError`,
 *  `AuthenticationError`, `NotFoundError` and `UnknownError`
 *  from the  {@link Response}'s parsed data
 */
var ResponseError = exports.ResponseError = /*#__PURE__*/function () {
  function ResponseError(response) {
    _classCallCheck(this, ResponseError);
    this.response = response;
    this.determineDescription();
  }

  // PRIVATE
  return _createClass(ResponseError, [{
    key: "determineDescription",
    value: function determineDescription() {
      if (!this.response || !this.response.parsed) {
        this.description = null;
        return;
      }
      var result = this.response.result;
      if (result && result.errors) {
        this.description = result.errors;
      } else if (result) {
        this.description = result;
      }
      return;
    }
  }]);
}(); // Protected
var NetworkError = exports.NetworkError = /*#__PURE__*/function (_ResponseError) {
  function NetworkError() {
    var _this;
    _classCallCheck(this, NetworkError);
    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }
    _this = _callSuper(this, NetworkError, [].concat(args));
    _this.code = 'NetworkError';
    return _this;
  }
  _inherits(NetworkError, _ResponseError);
  return _createClass(NetworkError);
}(ResponseError);
var ParserError = exports.ParserError = /*#__PURE__*/function (_ResponseError2) {
  function ParserError() {
    var _this2;
    _classCallCheck(this, ParserError);
    for (var _len2 = arguments.length, args = new Array(_len2), _key2 = 0; _key2 < _len2; _key2++) {
      args[_key2] = arguments[_key2];
    }
    _this2 = _callSuper(this, ParserError, [].concat(args));
    _this2.code = 'ParserError';
    return _this2;
  }
  _inherits(ParserError, _ResponseError2);
  return _createClass(ParserError);
}(ResponseError);
var ServerError = exports.ServerError = /*#__PURE__*/function (_ResponseError3) {
  function ServerError() {
    var _this3;
    _classCallCheck(this, ServerError);
    for (var _len3 = arguments.length, args = new Array(_len3), _key3 = 0; _key3 < _len3; _key3++) {
      args[_key3] = arguments[_key3];
    }
    _this3 = _callSuper(this, ServerError, [].concat(args));
    _this3.code = 'ServerError';
    return _this3;
  }
  _inherits(ServerError, _ResponseError3);
  return _createClass(ServerError);
}(ResponseError);
var ClientError = exports.ClientError = /*#__PURE__*/function (_ResponseError4) {
  function ClientError() {
    var _this4;
    _classCallCheck(this, ClientError);
    for (var _len4 = arguments.length, args = new Array(_len4), _key4 = 0; _key4 < _len4; _key4++) {
      args[_key4] = arguments[_key4];
    }
    _this4 = _callSuper(this, ClientError, [].concat(args));
    _this4.code = 'ClientError';
    return _this4;
  }
  _inherits(ClientError, _ResponseError4);
  return _createClass(ClientError);
}(ResponseError);
var AuthenticationError = exports.AuthenticationError = /*#__PURE__*/function (_ResponseError5) {
  function AuthenticationError() {
    var _this5;
    _classCallCheck(this, AuthenticationError);
    for (var _len5 = arguments.length, args = new Array(_len5), _key5 = 0; _key5 < _len5; _key5++) {
      args[_key5] = arguments[_key5];
    }
    _this5 = _callSuper(this, AuthenticationError, [].concat(args));
    _this5.code = 'AuthenticationError';
    return _this5;
  }
  _inherits(AuthenticationError, _ResponseError5);
  return _createClass(AuthenticationError);
}(ResponseError);
var NotFoundError = exports.NotFoundError = /*#__PURE__*/function (_ResponseError6) {
  function NotFoundError() {
    var _this6;
    _classCallCheck(this, NotFoundError);
    for (var _len6 = arguments.length, args = new Array(_len6), _key6 = 0; _key6 < _len6; _key6++) {
      args[_key6] = arguments[_key6];
    }
    _this6 = _callSuper(this, NotFoundError, [].concat(args));
    _this6.code = 'NotFoundError';
    return _this6;
  }
  _inherits(NotFoundError, _ResponseError6);
  return _createClass(NotFoundError);
}(ResponseError);
var UnknownError = exports.UnknownError = /*#__PURE__*/function (_ResponseError7) {
  function UnknownError() {
    var _this7;
    _classCallCheck(this, UnknownError);
    for (var _len7 = arguments.length, args = new Array(_len7), _key7 = 0; _key7 < _len7; _key7++) {
      args[_key7] = arguments[_key7];
    }
    _this7 = _callSuper(this, UnknownError, [].concat(args));
    _this7.code = 'UnknownError';
    return _this7;
  }
  _inherits(UnknownError, _ResponseError7);
  return _createClass(UnknownError);
}(ResponseError);
